### PR TITLE
(COMPLETED) Send Reminders To Email

### LINKED ISSUE #103: Send reminders to customers for appointments
Implement a way to send reminders to customers for all of their appointments

TODOs:
- [x] Create reminders when an appointment is created OR right after it starts (up to you, just note that if it is done on create, then the reminder should change if the appointment is cancelled)
- [x] Set the reminder types (one reminder for service, another for appointment confirmation)
- [x] Set the corresponding date (an arbitrary future date for the service reminder, and immediately for the confirmation)
- [x] Find a way to trigger the sending of the reminders booked for future dates (could be automated call once a day to check through all existing reminders)
- [x] Send the reminder to the customer's email

### PR DESCRIPTION
**Pull request by sandyl289**
## Branch: `sendReminder2Email`

> I don't know if everything will work like how I want them to and I also don't know how to verify if the methods I wrote work. If someone could help me, that would be great! 🥺

### Inside `EmailService.java`:
- [sendConfirmationEmail()](https://github.com/McGill-ECSE321-Winter2021/project-group-09/blob/bf3c7716fdf3a543c1f88080088987c6eb345b96/RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/utilities/EmailService.java#L29-L51): Sends a successful booking confirmation email to the recipient. 

- [x] We'll have to add this method in `createAppointment()`. When the user books an appointment, a confirmation email will be sent immediately

- [appointmentCancelled()](https://github.com/McGill-ECSE321-Winter2021/project-group-09/blob/bf3c7716fdf3a543c1f88080088987c6eb345b96/RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/utilities/EmailService.java#L88-L106): Sends a cancel confirmation email.

- [x] We'll have to add this method in `cancelAppointment()` (by admin&customer). 

- [x] Inside `cancelAppointment()`, we'll have to remove the related upcoming reminders for the cancelled appointment

- [sendReminderEmail()](https://github.com/McGill-ECSE321-Winter2021/project-group-09/blob/bf3c7716fdf3a543c1f88080088987c6eb345b96/RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/utilities/EmailService.java#L54-L86): Sends an upcomingAppointment reminder or serviceReminder. 

> This method will be called by the system automatically each day at 6 am (see [ScheduledTask.java](https://github.com/McGill-ECSE321-Winter2021/project-group-09/blob/bf3c7716fdf3a543c1f88080088987c6eb345b96/RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/utilities/ScheduledTasks.java#L68)) 

- [x] We'll have to call this method inside `createAppointment()` (Add future reminders `ReminderType.UpcomingAppointment`)
- [x] We'll have to call this method again inside `createAppointment()` to add a service reminder for the same service (Add service reminder `ReminderType.ServiceReminder`)
- [x] Inside `createAppointment()`,  we'll have to check if there's already a reminder of this type of service for the customer. If yes, replace previous reminder's dateTime by new dateTime.

### Inside `RepairShopApplication.java`:
- I added [@EnableScheduling](https://github.com/McGill-ECSE321-Winter2021/project-group-09/blob/bf3c7716fdf3a543c1f88080088987c6eb345b96/RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/RepairShopApplication.java#L12)

### Inside `ScheduledTasks.java`:
- [sendAllTodayReminder()](https://github.com/McGill-ECSE321-Winter2021/project-group-09/blob/bf3c7716fdf3a543c1f88080088987c6eb345b96/RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/utilities/ScheduledTasks.java#L38-L71):    
     Everyday at 6 am, this method will be called to send all today's `Service Reminders` & `Upcoming Appointment` reminders

### CODE REVIEW

**Review by Shaswata79 — APPROVED** [review]

File: RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/utilities/ScheduledTasks.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * Everyday at 6 am, this method will be called to send all today's:<br/>
+     * - Service Reminders<br/>
+     * - Upcoming Appointment
+     */
+    @Scheduled(cron = "0 0 6 * * ?")
+    public void sendAllTodayReminder() {
+
+        String today = systemDate();
```
Comment (T-Syme) [inline RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/utilities/ScheduledTasks.java]: To make testing easier, `systemDate()` should probably have a testing mode too
(Not urgent, just a note for future testing)

**Review by T-Syme — APPROVED** [review]

File: RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/AppointmentService.java [hunk 2]
```diff
@@ -39,6 +41,9 @@
     @Autowired
     BusinessRepository businessRepository;
 
+    @Autowired
+    private ReminderService reminderService;
```
Comment (JDziewonski98) [inline RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/AppointmentService.java]: autowiring a service in another service should be fine as long as it is 1 way: if you also autowire appointmentService in reminderService, it's a circular dependency

**Comment by JDziewonski98** [comment]
You can test the email method by just creating a temporary controller and passing your email to it, calling it postman and checking your email. That's what I did. Although if you just used the same logic as in my other method it should be fine. As for testing the scheduled method at 6am, maybe change it to 2 minutes after whatever time it is right now, run the app and wait lol?

**Review by JDziewonski98 — APPROVED** [review]
Looks good to me as long as the service methods are tested. I don't think we need to worry about writing tests for utilities / services that arent related to domain directly

**Pull request merged**